icelake/src/io/upsert_writer.rs

```diff
@@ -0,0 +1,221 @@
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    sync::Arc,
+};
+
+use crate::{
+    types::{Any, PartitionKey},
+    ErrorKind, Result,
+};
+use arrow_array::{Int32Array, RecordBatch};
+use arrow_schema::SchemaRef;
+
+use crate::{
+    config::TableConfigRef,
+    types::{PartitionSplitter, TableMetadata},
+    Error,
+};
+
+use super::{
+    file_writer::{DeltaWriterResult, EqualityDeltaWriter},
+    location_generator::FileLocationGenerator,
+    DefaultFileAppender, FileAppenderBuilder, FileAppenderLayer,
+};
+use arrow_ord::partition::partition;
+
+pub enum UpsertWriter<L: FileAppenderLayer<DefaultFileAppender>> {
```

Comment:
Is it possible to abstract this using generics? e.g. sth like
```
UpsertWriter<P: Partitioner, L: FileAppender>>
```

Reply:
Do you mean the user can specify partitioned or unpartitioned by `P: Partitioner`? Or different PartitionSplitter?

Reply:
> Do you mean the user can specify partitioned or unpartitioned by P: Partitioner? 

I mean this. Unpartitioned is just a special partitioning which always returns same partition key. I'm just guessing with this approach we can reduce some duplicated codes. But it's an optimization, and up to you on how to implement it.

Reply:
Oh, I get it. But for now, we directly use EqualityDeltaWriter as an Unpartitioned writer. So it didn't have duplicated codes between partitioned and unpartitioned writer.